src/gtimelog/main.py

```diff
@@ -794,35 +941,47 @@ def update_already_sent_indication(self, *args):
         else:
             self.infobar.hide()
 
-    def download_tasks(self):
+    def cancel_tasks_download(self, hide=True):
         if self._download:
-            return
+            old_message, old_url = self._download
+            soup_session.cancel_message(old_message, Soup.Status.CANCELLED)
+            self._download = None
+        if hide:
+            self.tasks_infobar.hide()
+
+    def download_tasks(self):
+        # hide=False and queue_resize() are needed to work around
+        # this bug: https://github.com/gtimelog/gtimelog/issues/89
+        self.cancel_tasks_download(hide=False)
+
         url = self.gsettings.get_string('task-list-url')
         if not url:
             log.debug("Not downloading tasks: URL not specified")
             return
         cache_filename = Settings().get_task_list_cache_file()
         self.tasks_infobar.set_message_type(Gtk.MessageType.INFO)
         self.tasks_infobar_label.set_text(_("Downloading tasks..."))
+        self.tasks_infobar.connect('response', lambda *args: self.cancel_tasks_download())
         self.tasks_infobar.show()
+        self.tasks_infobar.queue_resize()
         log.debug("Downloading tasks from %s", url)
-        cancellable = Gio.Cancellable()
-        gfile = Gio.File.new_for_uri(url)
-        self._download = (gfile, cancellable, url)
-        gfile.load_contents_async(cancellable, self.tasks_downloaded, cache_filename)
+        message = Soup.Message.new('GET', url)
+        self._download = (message, url)
+        soup_session.queue_message(message, self.tasks_downloaded, cache_filename)
 
-    def tasks_downloaded(self, source, result, cache_filename):
-        try:
-            success, content, etag = source.load_contents_finish(result)
-        except GLib.GError as e:
-            url = self._download[2]
-            log.error("Failed to download tasks from %s: %s", url, e)
+    def tasks_downloaded(self, session, message, cache_filename):
+        content = message.response_body.data
+        if message.status_code != Soup.Status.OK:
+            url = message.get_uri().to_string(just_path_and_query=False)
+
+            log.error("Failed to download tasks from %s: %d %s", url, message.status_code, message.reason_phrase)
             self.tasks_infobar.set_message_type(Gtk.MessageType.ERROR)
             self.tasks_infobar_label.set_text(_("Download failed."))
+            self.tasks_infobar.connect('response', lambda *args: self.infobar.hide())
             self.tasks_infobar.show()
         else:
-            log.debug("Successfully downloaded tasks (etag: %s):\n  %s",
-                      etag, content.decode('UTF-8', 'replace').replace('\n', '\n  '))
+            log.debug("Successfully downloaded tasks:\n  %s",
+                      content.decode('UTF-8', 'replace').replace('\n', '\n  '))
```

Comment:
I've recently tested this feature and discovered that this line of code throws an exception:

    AttributeError: 'str' object has no attribute 'decode'

because `content` returned by libsoap is already decoded into a unicode string, rather than a byte string.

I'm wondering if this is dependent on software versions (older libsoap?  older pygobject?  python 2 versus python 3)?  I have tested my fix (which is basically dropping the .decode() call) on both Python 2 and Python 3 and it seems to work fine there, but I'm not sure how to test older pygobject/libsoap versions?  Or was it always broken?

@andrewshadura do by any chance you remember if this worked for you?

Reply:
I almost forgot -- I also worry that the type of the returned data might depend on the Content-Type header returned by the HTTP server.